File: cs/Cifrado.cs
using System;
using System.IO;
using TTCrypto;

internal static class Cifrado
{
    private static int Main(string[] args)
    {
        try
        {
            if (args.Length != 4 && args.Length != 7 && args.Length != 6 && args.Length != 9)
            {
                Console.Error.WriteLine("Uso auto PNG/BMP/TIFF: Cifrado <in.png|in.bmp|in.tif|in.tiff> <out_cipher_u16.bin> <out_preview.png|bmp|tif|tiff|raw> <rondas>");
                Console.Error.WriteLine("Uso auto compartido: Cifrado <in.png|in.bmp|in.tif|in.tiff> <out_cipher_u16.bin> <out_preview.png|bmp|tif|tiff|raw> <rondas> <Z_hex_64> <salt_hex_64>");
                Console.Error.WriteLine("Uso RAW: Cifrado <in.raw> <out_cipher_u16.bin> <out_preview.raw|png|bmp|tif|tiff> <ancho> <alto> <canales> <rondas>");
                Console.Error.WriteLine("Uso RAW compartido: Cifrado <in.raw> <out_cipher_u16.bin> <out_preview.raw|png|bmp|tif|tiff> <ancho> <alto> <canales> <rondas> <Z_hex_64> <salt_hex_64>");
                return 1;
            }

            string inputPath = args[0];
            string outCipher = args[1];
            string outPreview = args[2];
            ushort rounds;
            uint width = 0, height = 0;
            ushort channels = 0;
            bool auto = args.Length == 4 || args.Length == 6;
            byte[]? z = null;
            byte[]? salt = null;

            if (auto)
            {
                rounds = ushort.Parse(args[3]);
                if (args.Length == 6)
                {
                    z = HexUtil.FromHex(args[4]);
                    salt = HexUtil.FromHex(args[5]);
                }
            }
            else
            {
                width = uint.Parse(args[3]);
                height = uint.Parse(args[4]);
                channels = ushort.Parse(args[5]);
                rounds = ushort.Parse(args[6]);
                if (args.Length == 9)
                {
                    z = HexUtil.FromHex(args[7]);
                    salt = HexUtil.FromHex(args[8]);
                }
            }

            var input = ImageUtil.LoadImageOrRaw(inputPath, width, height, channels);
            var ctx = new RcaContext(input.Height, input.Width, input.Channels, rounds);

            if (z is null || salt is null)
            {
                var (privA, pubA) = Keys.GenerateX25519Pair();
                var (privB, pubB) = Keys.GenerateX25519Pair();
                var zA = Keys.SharedSecretFromPkcs8(privA, pubB);
                var zB = Keys.SharedSecretFromPkcs8(privB, pubA);
                if (!zA.AsSpan().SequenceEqual(zB))
                    throw new InvalidOperationException("No se pudo acordar el secreto compartido.");
                z = zA;
                salt = Keys.GenerateSalt();
            }

            var session = Keys.DeriveSession(z, salt, ctx);

            var elems = input.Raw.Length;
            var prev = new ushort[elems];
            var cur = new ushort[elems];
            var curPerm = new ushort[elems];
            var next = new ushort[elems];

            for (var i = 0; i < elems; i++)
                cur[i] = (ushort)((input.Raw[i] + 1) % CryptoConsts.AutomataModBase);

            for (uint round = 0; round < rounds; round++)
            {
                var rk = Keys.DeriveRound(session, ctx, round);
                Array.Copy(cur, curPerm, elems);
                Permutations.ApplyU16(curPerm, input.Height, input.Width, input.Channels, rk.PermIndex);
                var k1 = Automata.KernelFromMoore8(rk.Moore1);
                var k2 = Automata.KernelFromMoore8(rk.Moore2);
                Automata.StepU16(prev, curPerm, next, input.Height, input.Width, input.Channels, k1, k2, round, BoundaryHelper.FromRound(rk));
                (prev, cur, next) = (cur, next, prev);
            }

            WriteU16(outCipher, cur);
            var prevPath = outCipher + ".prev.bin";
            WriteU16(prevPath, prev);

            var preview = new byte[cur.Length];
            for (var i = 0; i < cur.Length; i++)
                preview[i] = (byte)Math.Min(cur[i], (ushort)255);
            ImageUtil.SavePreview(preview, input.Width, input.Height, input.Channels, outPreview);

            var sessionFile = new SessionFile
            {
                Width = input.Width.ToString(),
                Height = input.Height.ToString(),
                Channels = input.Channels.ToString(),
                Rounds = rounds.ToString(),
                PrevPath = prevPath,
                CurPath = outCipher,
                ZHex = HexUtil.ToHex(z),
                SaltHex = HexUtil.ToHex(salt)
            };
            sessionFile.Save(outCipher + ".session.txt");

            Console.WriteLine("Cifrado completado.");
            Console.WriteLine($"Entrada: {inputPath}");
            Console.WriteLine($"Salida estado mod257 (u16): {outCipher}");
            Console.WriteLine($"Salida estado previo (u16): {prevPath}");
            Console.WriteLine($"Archivo de sesion: {outCipher}.session.txt");
            Console.WriteLine($"Salida preview recortada (u8): {outPreview}");
            Console.WriteLine($"Dimensiones: {input.Width}x{input.Height} x {input.Channels}, rondas={rounds}");
            return 0;
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine(ex.Message);
            return 2;
        }
    }

    private static void WriteU16(string path, ushort[] data)
    {
        using var bw = new BinaryWriter(File.Create(path));
        foreach (var v in data) bw.Write(v);
    }
}

File: java/Cifrado.java
import java.nio.file.Path;
import java.util.HexFormat;

public final class Cifrado {
    public static void main(String[] args) throws Exception {
        if (args.length != 4 && args.length != 7 && args.length != 6 && args.length != 9) {
            System.err.println("Uso auto PNG/BMP/TIFF: java Cifrado <in.png|in.bmp|in.tif|in.tiff> <out_cipher_u16.bin> <out_preview.png|bmp|tif|tiff|raw> <rondas>");
            System.err.println("Uso auto compartido: java Cifrado <in.png|in.bmp|in.tif|in.tiff> <out_cipher_u16.bin> <out_preview.png|bmp|tif|tiff|raw> <rondas> <Z_hex_64> <salt_hex_64>");
            System.err.println("Uso RAW: java Cifrado <in.raw> <out_cipher_u16.bin> <out_preview.raw|png|bmp|tif|tiff> <ancho> <alto> <canales> <rondas>");
            System.err.println("Uso RAW compartido: java Cifrado <in.raw> <out_cipher_u16.bin> <out_preview.raw|png|bmp|tif|tiff> <ancho> <alto> <canales> <rondas> <Z_hex_64> <salt_hex_64>");
            System.exit(1);
        }

        String inputPath = args[0];
        String outCipher = args[1];
        String outPreview = args[2];
        int rounds;
        Integer width = null, height = null, channels = null;
        byte[] z = null;
        byte[] salt = null;

        if (args.length == 4 || args.length == 6) {
            rounds = Integer.parseInt(args[3]);
            if (args.length == 6) {
                z = HexFormat.of().parseHex(args[4]);
                salt = HexFormat.of().parseHex(args[5]);
            }
        } else {
            width = Integer.parseInt(args[3]);
            height = Integer.parseInt(args[4]);
            channels = Integer.parseInt(args[5]);
            rounds = Integer.parseInt(args[6]);
            if (args.length == 9) {
                z = HexFormat.of().parseHex(args[7]);
                salt = HexFormat.of().parseHex(args[8]);
            }
        }

        Common.LoadedImage input = Common.loadImageOrRaw(inputPath, width, height, channels);
        Common.RcaContext ctx = new Common.RcaContext(input.height(), input.width(), input.channels(), rounds);
        if (z == null || salt == null) {
            z = Llaves.deriveSharedSecret();
            salt = Llaves.generateSalt();
        }
        var session = Llaves.deriveSession(z, salt, ctx);

        int elems = input.raw().length;
        short[] prev = new short[elems];
        short[] cur = new short[elems];
        short[] curPerm = new short[elems];
        short[] next = new short[elems];

        for (int i = 0; i < elems; i++) {
            cur[i] = (short)((Byte.toUnsignedInt(input.raw()[i]) + 1) % Common.AUTOMATA_MOD_BASE);
        }

        for (int round = 0; round < rounds; round++) {
            var rk = Llaves.deriveRound(session, ctx, round);
            System.arraycopy(cur, 0, curPerm, 0, elems);
            Permutaciones.applyU16(curPerm, input.height(), input.width(), input.channels(), rk.permIndex());
            int[][][][] kernels = new int[3][2][][];
            for (int ch = 0; ch < 3; ch++) {
                for (int k = 0; k < 2; k++) {
                    kernels[ch][k] = Automata.kernelFromMoore8(rk.mooreByChannel()[ch][k]);
                }
            }
            Automata.stepU16(prev, curPerm, next, input.height(), input.width(), input.channels(), kernels, round, Llaves.boundaryFromRound(rk));
            short[] tmp = prev; prev = cur; cur = next; next = tmp;
        }

        Common.writeU16(Path.of(outCipher), cur);
        String prevPath = outCipher + ".prev.bin";
        Common.writeU16(Path.of(prevPath), prev);

        byte[] preview = new byte[cur.length];
        for (int i = 0; i < cur.length; i++) preview[i] = (byte)Math.min(Short.toUnsignedInt(cur[i]), 255);
        Common.savePreview(preview, input.width(), input.height(), input.channels(), outPreview);

        Common.SessionFile sessionFile = new Common.SessionFile();
        sessionFile.width = Integer.toString(input.width());
        sessionFile.height = Integer.toString(input.height());
        sessionFile.channels = Integer.toString(input.channels());
        sessionFile.rounds = Integer.toString(rounds);
        sessionFile.prevPath = prevPath;
        sessionFile.curPath = outCipher;
        sessionFile.zHex = HexFormat.of().formatHex(z);
        sessionFile.saltHex = HexFormat.of().formatHex(salt);
        sessionFile.save(Path.of(outCipher + ".session.txt"));

        System.out.println("Cifrado completado.");
        System.out.println("Entrada: " + inputPath);
        System.out.println("Salida estado mod257 (u16): " + outCipher);
        System.out.println("Salida estado previo (u16): " + prevPath);
        System.out.println("Archivo de sesion: " + outCipher + ".session.txt");
        System.out.println("Salida preview recortada (u8): " + outPreview);
        System.out.printf("Dimensiones: %dx%d x %d, rondas=%d%n", input.width(), input.height(), input.channels(), rounds);
    }
}
